File: typescript/src/oauth.ts
/**
 * OAuth 2.0 Authorization Code Flow with PKCE for the Surf API.
 *
 * @example
 * ```ts
 * import { SurfOAuth } from '@surf/api/oauth';
 *
 * const oauth = new SurfOAuth({
 *   clientId: 'your-app-public-id',
 *   redirectUri: 'https://yourapp.com/callback',
 * });
 *
 * // Step 1: Generate authorization URL
 * const { url, state, codeVerifier } = oauth.getAuthorizeUrl({ scope: 'read:feeds write:statuses' });
 *
 * // Step 2: Exchange code after redirect
 * const tokens = await oauth.exchangeCode(code, codeVerifier);
 *
 * // Step 3: Use the access token
 * const client = new SurfClient({ apiKey: tokens.access_token });
 *
 * // Step 4: Refresh when expired
 * const newTokens = await oauth.refreshToken(tokens.refresh_token);
 * ```
 */

const DEFAULT_AUTH_URL = 'https://surf.social';
const DEFAULT_TOKEN_URL = 'https://surf.social/oauth/token';
const DEFAULT_REVOKE_URL = 'https://surf.social/oauth/revoke';

export interface SurfOAuthOptions {
  clientId: string;
  redirectUri: string;
  authBaseUrl?: string;
  tokenUrl?: string;
  revokeUrl?: string;
}

export interface AuthorizeUrlResult {
  url: string;
  state: string;
  codeVerifier: string;
}

export interface OAuthTokens {
  access_token: string;
  token_type: string;
  expires_in: number;
  refresh_token: string;
  scope: string;
}

/**
 * Generate a cryptographically secure random string.
 */
function randomString(length: number): string {
  const array = new Uint8Array(length);
  crypto.getRandomValues(array);
  return Array.from(array, (b) => b.toString(16).padStart(2, '0')).join('');
}

/**
 * Generate PKCE code_verifier and code_challenge (S256).
 */
export async function generatePKCE(): Promise<{ verifier: string; challenge: string }> {
  const verifier = randomString(64);
  const encoder = new TextEncoder();
  const data = encoder.encode(verifier);
  const hash = await crypto.subtle.digest('SHA-256', data);
  const challenge = btoa(String.fromCharCode(...new Uint8Array(hash)))
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '');
  return { verifier, challenge };
}

export class SurfOAuth {
  private readonly clientId: string;
  private readonly redirectUri: string;
  private readonly authBaseUrl: string;
  private readonly tokenUrl: string;
  private readonly revokeUrl: string;

  constructor(options: SurfOAuthOptions) {
    this.clientId = options.clientId;
    this.redirectUri = options.redirectUri;
    this.authBaseUrl = (options.authBaseUrl ?? DEFAULT_AUTH_URL).replace(/\/+$/, '');
    this.tokenUrl = options.tokenUrl ?? DEFAULT_TOKEN_URL;
    this.revokeUrl = options.revokeUrl ?? DEFAULT_REVOKE_URL;
  }

  /**
   * Build the authorization URL with PKCE.
   * Returns the URL to redirect the user to, plus state and verifier to save.
   */
  async getAuthorizeUrl(options?: {
    scope?: string;
    state?: string;
  }): Promise<AuthorizeUrlResult> {
    const state = options?.state ?? randomString(32);
    const { verifier, challenge } = await generatePKCE();

    const params = new URLSearchParams({
      client_id: this.clientId,
      redirect_uri: this.redirectUri,
      scope: options?.scope ?? 'read:feeds',
      response_type: 'code',
      state,
      code_challenge: challenge,
      code_challenge_method: 'S256',
    });

    return {
      url: `${this.authBaseUrl}/oauth/authorize?${params.toString()}`,
      state,
      codeVerifier: verifier,
    };
  }

  /**
   * Exchange an authorization code for access + refresh tokens.
   */
  async exchangeCode(code: string, codeVerifier: string): Promise<OAuthTokens> {
    const resp = await fetch(this.tokenUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        grant_type: 'authorization_code',
        client_id: this.clientId,
        code,
        redirect_uri: this.redirectUri,
        code_verifier: codeVerifier,
      }),
    });
    if (!resp.ok) {
      const err = await resp.json().catch(() => ({}));
      throw new Error(err.error_description ?? err.error ?? `Token exchange failed: ${resp.status}`);
    }
    return resp.json();
  }

  /**
   * Refresh an access token using a refresh token.
   * Note: refresh tokens rotate — the old one is invalidated.
   */
  async refreshToken(refreshToken: string): Promise<OAuthTokens> {
    const resp = await fetch(this.tokenUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        grant_type: 'refresh_token',
        client_id: this.clientId,
        refresh_token: refreshToken,
      }),
    });
    if (!resp.ok) {
      const err = await resp.json().catch(() => ({}));
      throw new Error(err.error_description ?? err.error ?? `Token refresh failed: ${resp.status}`);
    }
    return resp.json();
  }

  /**
   * Revoke an access or refresh token.
   */
  async revoke(token: string): Promise<boolean> {
    const resp = await fetch(this.revokeUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        client_id: this.clientId,
        token,
      }),
    });
    return resp.ok;
  }
}

File: go/oauth.go
// Package surf provides OAuth 2.0 Authorization Code Flow with PKCE helpers.
//
// Usage:
//
//	oauth := surf.NewOAuth("client-id", "https://yourapp.com/callback")
//	url, state, verifier := oauth.GetAuthorizeURL("read:feeds write:statuses", "")
//	// redirect user to url, save state + verifier
//	tokens, err := oauth.ExchangeCode(code, verifier)
//	// use tokens.AccessToken with surf.NewClient()
package surf

import (
	"bytes"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"
)

const (
	defaultAuthBaseURL = "https://surf.social"
	defaultTokenURL    = "https://surf.social/oauth/token"
	defaultRevokeURL   = "https://surf.social/oauth/revoke"
)

// OAuthTokens holds the tokens returned from the authorization server.
type OAuthTokens struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int    `json:"expires_in"`
	RefreshToken string `json:"refresh_token"`
	Scope        string `json:"scope"`
}

// OAuth is a helper for the Surf OAuth 2.0 Authorization Code Flow with PKCE.
type OAuth struct {
	ClientID    string
	RedirectURI string
	AuthBaseURL string
	TokenURL    string
	RevokeURL   string
	HTTP        *http.Client
}

// NewOAuth creates a new OAuth helper with default URLs.
func NewOAuth(clientID, redirectURI string) *OAuth {
	return &OAuth{
		ClientID:    clientID,
		RedirectURI: redirectURI,
		AuthBaseURL: defaultAuthBaseURL,
		TokenURL:    defaultTokenURL,
		RevokeURL:   defaultRevokeURL,
		HTTP:        &http.Client{Timeout: 15 * time.Second},
	}
}

// GeneratePKCE generates a code_verifier and code_challenge (S256).
func GeneratePKCE() (verifier, challenge string) {
	b := make([]byte, 64)
	rand.Read(b)
	verifier = hex.EncodeToString(b)
	h := sha256.Sum256([]byte(verifier))
	challenge = base64.RawURLEncoding.EncodeToString(h[:])
	return
}

// GetAuthorizeURL builds the authorization URL with PKCE.
// If state is empty, a random state is generated.
// Returns (url, state, codeVerifier).
func (o *OAuth) GetAuthorizeURL(scope, state string) (string, string, string) {
	if state == "" {
		b := make([]byte, 32)
		rand.Read(b)
		state = hex.EncodeToString(b)
	}
	verifier, challenge := GeneratePKCE()

	params := url.Values{
		"client_id":             {o.ClientID},
		"redirect_uri":         {o.RedirectURI},
		"scope":                {scope},
		"response_type":        {"code"},
		"state":                {state},
		"code_challenge":       {challenge},
		"code_challenge_method": {"S256"},
	}
	return fmt.Sprintf("%s/oauth/authorize?%s", o.AuthBaseURL, params.Encode()), state, verifier
}

// ExchangeCode exchanges an authorization code for tokens.
func (o *OAuth) ExchangeCode(code, codeVerifier string) (*OAuthTokens, error) {
	body, err := json.Marshal(map[string]string{
		"grant_type":    "authorization_code",
		"client_id":     o.ClientID,
		"code":          code,
		"redirect_uri":  o.RedirectURI,
		"code_verifier": codeVerifier,
	})
	if err != nil {
		return nil, fmt.Errorf("marshal token request: %w", err)
	}
	resp, err := o.HTTP.Post(o.TokenURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read token response: %w", err)
	}
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("token exchange failed (%d): %s", resp.StatusCode, string(data))
	}
	var tokens OAuthTokens
	if err := json.Unmarshal(data, &tokens); err != nil {
		return nil, err
	}
	return &tokens, nil
}

// RefreshToken uses a refresh token to get new tokens.
// Note: refresh tokens rotate — the old one is invalidated.
func (o *OAuth) RefreshToken(refreshToken string) (*OAuthTokens, error) {
	body, err := json.Marshal(map[string]string{
		"grant_type":    "refresh_token",
		"client_id":     o.ClientID,
		"refresh_token": refreshToken,
	})
	if err != nil {
		return nil, fmt.Errorf("marshal refresh request: %w", err)
	}
	resp, err := o.HTTP.Post(o.TokenURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("read refresh response: %w", err)
	}
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("token refresh failed (%d): %s", resp.StatusCode, string(data))
	}
	var tokens OAuthTokens
	if err := json.Unmarshal(data, &tokens); err != nil {
		return nil, err
	}
	return &tokens, nil
}

// Revoke revokes an access or refresh token.
func (o *OAuth) Revoke(token string) error {
	body, err := json.Marshal(map[string]string{
		"client_id": o.ClientID,
		"token":     token,
	})
	if err != nil {
		return fmt.Errorf("marshal revoke request: %w", err)
	}
	resp, err := o.HTTP.Post(o.RevokeURL, "application/json", bytes.NewReader(body))
	if err != nil {
		return err
	}
	resp.Body.Close()
	if resp.StatusCode != 200 {
		return fmt.Errorf("revocation failed: %d", resp.StatusCode)
	}
	return nil
}
